File: hyprbar/modules/power_menu.py
from gi.repository import Gtk
from hyprbar.widgets.module_container import ModuleContainer
from hyprbar.common import run_command

OPTIONS = [
    {
        "title": "Shutdown",
        "icon": "system-shutdown-symbolic",
        "command": "shutdown now",
    },
    {"title": "Reboot", "icon": "system-reboot-symbolic", "command": "reboot"},
    {
        "title": "Logout",
        "icon": "system-log-out-symbolic",
        "command": "hyprctl dispatch exit",
    },
    {
        "title": "Sleep",
        "icon": "weather-clear-night-symbolic",
        "command": "systemctl suspend",
    },
]


class PowerMenu(ModuleContainer):
    def __init__(self, config, window):
        super().__init__(config, window, is_button=True)
        shutdown_icon = Gtk.Image.new_from_icon_name(
            "system-shutdown-symbolic")
        box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)

        for option in OPTIONS:
            button = Gtk.Button()
            button.add_css_class("flat")
            button_box = Gtk.Box(
                orientation=Gtk.Orientation.HORIZONTAL, spacing=10)
            button_box.append(Gtk.Image.new_from_icon_name(option["icon"]))
            button_box.append(Gtk.Label(label=option["title"]))
            button.set_child(button_box)

            # command=option["command"] is needed
            # to sure all functions don't end up being the same
            def on_click(_, command=option["command"]):
                run_command(command)

            button.connect("clicked", on_click)
            box.append(button)

        self.set_child(shutdown_icon)

        self.popover = Gtk.Popover()
        self.popover.set_child(box)
        self.popover.set_position(Gtk.PositionType.BOTTOM)
        self.popover.set_has_arrow(False)
        self.popover.set_offset(-42, 8)
        self.popover.set_parent(self.container)
        self.container.connect("clicked", self.on_click)

    def on_click(self, window):
        self.popover.popup()

File: src/modules/power_menu.rs
use relm4::{gtk, ComponentParts, ComponentSender, SimpleComponent, RelmWidgetExt, WidgetTemplate};
use relm4::gtk::prelude::*;
use crate::common::execute_command;

#[derive(Debug)]
pub struct PowerMenu {
}

#[derive(Debug)]
pub enum PowerMenuInput {
    Shutdown,
    Restart,
    Sleep,
    Logout
}

#[relm4::widget_template(pub)]
impl WidgetTemplate for MenuButton {
    view! {
        gtk::Button {
            add_css_class: "flat",
            gtk::Box {
                set_orientation: gtk::Orientation::Horizontal,
                set_spacing: 10,
                #[name = "image"]
                gtk::Image {},
                #[name = "label"]
                gtk::Label {}
            }
        }

    }
}

#[relm4::component(pub)]
impl SimpleComponent for PowerMenu {
    type Input = PowerMenuInput;
    type Output = ();
    type Init = ();

    view!{
        gtk::Box {
            gtk::Button {
                inline_css: "background: transparent; padding: 0px 8px",
                gtk::Image::from_icon_name("system-shutdown-symbolic") {
                },
                connect_clicked[popover] => move |_| {
                    popover.popup();
                }

            },
            #[name = "popover"]
            gtk::Popover {
                set_css_classes: &["popover-padding"],
                set_has_arrow: false,
                gtk::Box {
                    set_orientation: gtk::Orientation::Vertical,
                    #[template]
                    MenuButton {
                        connect_clicked: |_| {
                            execute_command("shutdown now");
                        },
                        #[template_child]
                        image {
                            set_from_icon_name: Some("system-shutdown-symbolic"),
                        },
                        #[template_child]
                        label {
                            set_label: "Shutdown"
                        }
                    },
                    #[template]
                    MenuButton {
                        connect_clicked: |_| {
                            execute_command("reboot");
                        },
                        #[template_child]
                        image {
                            set_from_icon_name: Some("system-reboot-symbolic"),
                        },
                        #[template_child]
                        label {
                            set_label: "Reboot"
                        }
                    },
                    #[template]
                    MenuButton {
                        connect_clicked: |_| {
                            execute_command("hyprctl dispatch exit");
                        },
                        #[template_child]
                        image {
                            set_from_icon_name: Some("system-log-out-symbolic"),
                        },
                        #[template_child]
                        label {
                            set_label: "Logout"
                        }
                    },
                    #[template]
                    MenuButton {
                        connect_clicked: |_| {
                            execute_command("systemctl suspend");
                        },
                        #[template_child]
                        image {
                            set_from_icon_name: Some("weather-clear-night-symbolic"),
                        },
                        #[template_child]
                        label {
                            set_label: "Sleep"
                        }
                    }

                }
            }

        }
    } 

    /// Initialize the UI and model.
    fn init(
        _config: Self::Init,
        root_box: &Self::Root,
        _sender: ComponentSender<Self>,
    ) -> relm4::ComponentParts<Self> {
        let model = PowerMenu {
        };

        let widgets = view_output!();

        ComponentParts { model, widgets }
    }

    /// Update model based on message
    fn update(&mut self, message: Self::Input, _sender: ComponentSender<Self>) {
        match message {
            PowerMenuInput::Shutdown =>  {}
            PowerMenuInput::Restart =>  {}
            PowerMenuInput::Sleep =>  {}
            PowerMenuInput::Logout =>  {}
        }
    }
}
